plugins/inputs/stackdriver/stackdriver.go

```diff
@@ -0,0 +1,690 @@
+	tsChan := make(chan *monitoringpb.TimeSeries, 1000)
+
+	go func() {
+		// Channel must be closed for safety
+		defer close(tsChan)
+
+		// Iterate over timeseries and send them to buffered channel
+		tsResp := c.conn.ListTimeSeries(ctx, req)
+		for {
+			tsDesc, tsErr := tsResp.Next()
+			if tsErr != nil {
+				if tsErr != iterator.Done {
+					log.Printf("D! Request %s failure: %s\n", req.String(), tsErr)
+				}
+				break
+			}
+			tsChan <- tsDesc
+		}
+	}()
+
+	return tsChan, nil
+}
+
+// Close implements metricClient interface
+func (c *stackdriverMetricClient) Close() error {
+	return c.conn.Close()
+}
+
+func init() {
+	f := func() telegraf.Input {
+		return &Stackdriver{
+			RateLimit:                       14,
+			LookbackSeconds:                 120,
+			DelaySeconds:                    60,
+			ScrapeDistributionBuckets:       true,
+			DistributionAggregationAligners: []string{},
+		}
+	}
+
+	inputs.Add("stackdriver", f)
+}
+
+// Description implements telegraf.inputs interface
+func (s *Stackdriver) Description() string {
+	return description
+}
+
+// SampleConfig implements telegraf.inputs interface
+func (s *Stackdriver) SampleConfig() string {
+	return sampleConfig
+}
+
+// Gather implements telegraf.inputs interface
+func (s *Stackdriver) Gather(acc telegraf.Accumulator) error {
+	err := s.initializeStackdriverClient()
+	if err != nil {
+		log.Printf("E! Failed to create stackdriver monitoring client: %v", err)
+		return err
+	}
+
+	s.updateWindow()
+
+	tsConfs, err := s.generatetimeSeriesConfs()
+	if err != nil {
+		log.Printf("E! Failed to get metrics: %s\n", err)
+		return err
+	}
+
+	var wg sync.WaitGroup
+	lmtr := limiter.NewRateLimiter(s.RateLimit, time.Second)
+	defer lmtr.Stop()
+
+	wg.Add(len(tsConfs))
+	for _, tsConf := range tsConfs {
+		<-lmtr.C
+		go func(tsConf *timeSeriesConf) {
+			defer wg.Done()
+			acc.AddError(s.scrapeTimeSeries(acc, tsConf))
+		}(tsConf)
+	}
+	wg.Wait()
+
+	return nil
+}
+
+func (s *Stackdriver) updateWindow() {
+	windowEnd := time.Now().Add(-time.Duration(s.DelaySeconds) * time.Second)
+
+	if s.windowEnd.IsZero() {
+		s.windowStart = windowEnd.Add(-time.Duration(s.LookbackSeconds) * time.Second)
+	} else {
+		s.windowStart = s.windowEnd
+	}
+	s.windowEnd = windowEnd
+}
+
+// Internal structure which holds our configuration for a particular GCP time
+// series.
+type timeSeriesConf struct {
+	// The influx measurement name this time series maps to
+	measurement string
+	// The prefix to use before any influx field names that we'll write for
+	// this time series. (Or, if we only decide to write one field name, this
+	// field just holds the value of the field name.)
+	fieldPrefix string
+	// The GCP API request that we'll use to fetch data for this time series.
+	listTimeSeriesRequest *monitoringpb.ListTimeSeriesRequest
+}
+
+// Generate filter string for ListTimeSeriesRequest
+func (s *Stackdriver) newListTimeSeriesFilter(metricType string) string {
+	if s.Filter == nil {
+		return ""
```

Comment:
This is what is causing the "The provided filter matches more than one metric." error that I saw. The filterString defined in line 326 should be returned when the user hasn't defined any additional filters. We always have to filter by a single metric type.

After I made this change in https://github.com/sciurus/telegraf/commit/352a57902ff122b9267a054b9e855b475b6d062a I got metrics from stackdriver in my output.